### ISSUE TITLE
Using `isinstance(obj, A | B)` incorrectly requires generic parameters for classes

### ISSUE THREAD

**Issue by bryanforbes** [microsoft/pyright#5294]
**Describe the bug**
Given the following code:

```python
from __future__ import annotations

from typing import Any, Generic, TypeVar

T = TypeVar('T')


class Thing(Generic[T]):
    prop: T


class OtherThing:
    ...


def func(thing: Any) -> None:
    if isinstance(thing, Thing | OtherThing):
        ...
```

An error is reported at line 17:

```
Expected type arguments for generic class "Thing"
```

**Expected behavior**
No errors are reported

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, or the command-line tool?
Command line

Which version?
1.1.314

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by Hootner** [comment]
It seems that you are reporting that an error is being reported at line 17, although there shouldn't be one. The error message mentions "Expected type arguments for generic class 'Thing'". You are running Pyright version 1.1.314 as a command-line tool. I understand the issue and will try to help you out.

The error you are encountering is because you are missing the type argument for `Thing` at line 17. Since `Thing` is a generic class, it should take a type argument like `Thing[int]` or `Thing[str]` when used. The problem is occurring because `isinstance` is used with a union of both `Thing` and `OtherThing` as the second argument. Instead of using `Thing | OtherThing`, you need to replace it with the appropriate type argument for `Thing`.

One way to correct the code is to use a type variable for the type of `thing` when using the union:

```python
from __future__ import annotations

from typing import Any, Generic, TypeVar

T = TypeVar('T')
U = TypeVar('U', bound='Thing')  # Added a bound to specifically target the Thing class


class Thing(Generic[T]):
    prop: T


class OtherThing:
    ...


def func(thing: Any) -> None:
    if isinstance(thing, U | OtherThing):  # Replaced Thing with U
        ...
```

By adding the bound `U`, you tell the type checker that `U` should be an instance of a subclass or a subclass of an instance of `Thing`. This way, you preserve the generic nature of `Thing` and resolve the error.

**Comment by erictraut** [comment]
This is included in pyright 1.1.315, which I just published. It will also be included in this week's insiders release of pylance.

**Comment by mgmira** [comment]
I'm still being able to reproduce this in specific conditions, with latest version of pyright
```python
# pyright: strict

from __future__ import annotations

from typing import Any


class RestApi:
    @staticmethod
    def output_json(data: int | str | dict[str, Any]):
        if isinstance(data, dict | str):
            return data
```

**Comment by erictraut** [comment]
Yeah, sorry about that. There was a second (infrequent) code path that needed to be updated as well. It will be addressed in the next release.
